### PR TITLE
PLANNER-1613 toList() and toSet() constraint collectors

### PR DESCRIPTION
**Pull request by triceo**
This is a simple "last resort" collector for when our other collectors are insufficient. People can always retrieve a collection of the facts, and perform whatever operation they want on them afterwards.

WIthout providing this functionality, people would have to resort to submitting feature requests and be blocked before we implement them.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 1]
```diff
@@ -554,4 +561,171 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
     private ConstraintCollectors() {
     }
 
+    // ************************************************************************
+    // toCollection
+    // ************************************************************************
+
+    /**
+     * Creates constraint collector that will return matched facts as a collection.
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: avoid future tense.
-> "that returns"

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 2]
```diff
@@ -554,4 +561,171 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
...
+     *
+     * @param collectionFunction never null, provides new instance of the target collection with given initial capacity
+     * @param <A> type of the matched fact, matching the type of elements in the resulting collection
+     * @param <Result> type of the target collection
+     * @return never null
+     */
+    public static <A, Result extends Collection<A>> UniConstraintCollector<A, ?, Result> toCollection(
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: Do we need toCollection? When in doubt leave it out. I 'd rather start with just toSet() and toList() until someone complains of the lack of toCollection(). Wdyt?

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 3]
```diff
@@ -554,4 +561,171 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
...
+            IntFunction<Result> collectionFunction) {
+        return toCollection(Function.identity(), collectionFunction);
+    }
+
+    public static <A> UniConstraintCollector<A, ?, Set<A>> toSet() {
+        return toSet(Function.identity());
+    }
+
+    public static <A> UniConstraintCollector<A, ?, List<A>> toList() {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: javadocs :)

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 2 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: I consider it useful for when you want a stable iteration order, so you request `toCollection(TreeSet::new)`.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java [hunk 4]
```diff
@@ -554,4 +561,171 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
...
+                (resultContainer, a) -> {
+                    Mapped mapped = mappingFunction.apply(a);
+                    resultContainer.add(mapped);
+                    return () -> resultContainer.remove(mapped);
+                },
+                resultContainer -> finishToCollection(collectionFunction, resultContainer));
+    }
+
+    private static <Mapped, Container extends List<Mapped>, Result extends Collection<Mapped>> Result
+    finishToCollection(IntFunction<Result> collectionFunction, Container resultContainer) {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: Thinking aloud: The method name doesn't indicate it's used as the undo operation.
How about calling it to undoToCollection?

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 3 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: Yeah, the Javadoc is on `toCollection()` now.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 4 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: This is not the undo. This is the finisher, which takes the incremental container and turns it into a final immutable product.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: Agreed, then the name is correct. I interpreted "finish" as last part that is common, not as finisher.

Maybe `toCollectionFinisher`? Although name is fine now.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 3 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: Add at least a "As defined by ..."

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java (hunk 2 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: Agreed, or to pick your collection type. Keep it indeed.

This is the special case though. The toList() is the normal case. So toList() gets a full proper javadoc (they others can use "As defined by"). Wdyt?

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java]: So, we want `toCollection(constructor)`, but not `toCollection()`?

**Pull request merged**